### ISSUE TITLE
Timeout on getSubmissionDataForChart - to much data in query response

### ISSUE THREAD

**Issue by Tobi77** [mosparo/mosparo#121]
Hi,

we run a mosparo instance 0.4.3 and run into the following error after selecting the project and displaying the chart in the backend.
`Error: Maximum execution time of 30 seconds exceeded at vendor/paragonie/sodium_compat/src/Core/Util.php:583)"}`
We do have ~4000 submissions in the database.

The issue comes from `Controller/ProjectRelated/DashboardController.php:getSubmissionDataForChart` and selected the whole content from submissions. 

This request took 8.6 seconds to fetch.
```
SELECT s0_.id AS id_0, s0_.validation_token AS validation_token_1, s0_.data AS data_2, s0_.signature AS signature_3, s0_.submitted_at AS submitted_at_4, s0_.verified_at AS verified_at_5, s0_.matched_rule_items AS matched_rule_items_6, s0_.ignored_fields AS ignored_fields_7, s0_.verified_fields AS verified_fields_8, s0_.general_verifications AS general_verifications_9, s0_.spam_rating AS spam_rating_10, s0_.spam AS spam_11, s0_.spam_detection_rating AS spam_detection_rating_12, s0_.valid AS valid_13, s0_.submit_token_id AS submit_token_id_14, s0_.project_id AS project_id_15 
FROM submission s0_ 
WHERE (s0_.spam = 1 OR s0_.valid IS NOT NULL) 
AND (s0_.project_id = 1)
```

If we limit the select to the necessary fields we could speed up the request because we only need 3 fields to display the chart (submittedAt, valid, spam).

```
SELECT s0_.submitted_at AS submitted_at_4, s0_.valid AS valid_13, s0_.spam AS spam_11
FROM submission s0_ 
WHERE (s0_.spam = 1 OR s0_.valid IS NOT NULL) 
AND (s0_.project_id = 1)
```
0.028 Sec

```
protected function getSubmissionDataForChart($entityManager): array
    {
        $noSpamSubmissionsData = $spamSubmissionsData = $this->createEmptyDateArray();

        $builder = $entityManager->createQueryBuilder();
        $builder
            ->select('s.submittedAt, s.valid, s.spam')
            ->from(Submission::class, 's')
            ->where('s.spam = 1')
            ->orWhere('s.valid IS NOT NULL');

        foreach ($builder->getQuery()->getResult() as $submission) {
            $dateKey = $submission['submittedAt']->format('Y-m-d');
            if ($submission['spam'] || !$submission['valid']) {
                if (!isset($spamSubmissionsData[$dateKey])) {
                    continue;
                }

                $spamSubmissionsData[$dateKey]++;
            } else if ($submission['valid']) {
                if (!isset($noSpamSubmissionsData[$dateKey])) {
                    continue;
                }

                $noSpamSubmissionsData[$dateKey]++;
            }
        }

        return [
            $this->convertIntoChartArray($noSpamSubmissionsData),
            $this->convertIntoChartArray($spamSubmissionsData),
            array_sum($noSpamSubmissionsData),
            array_sum($spamSubmissionsData)
        ];
    }
```

A different approach could be to cache all data in the past and only fetch the current day from the database.

What do you think how should we address this issue?

**Comment by zepich** [comment]
Hi @Tobi77 

Thank you very much for your report and your suggestion.

I think your suggestion is perfect! I wonder why I didn't do it as you suggested in the first place since loading all the submissions makes no sense... :|

We already had the idea to add a table to store the statistics in there (date, number of valid submissions, number of invalid submissions) so that the dashboard could use these data and could show the data from the past (even weeks, months, or years ago, because these statistical data would not be deleted). But that's more work to do, and I think we will add that for one of the next major versions. When we do that, we can replace the data source for the chart on the dashboard with this table and allow the user to select the time range.

I will fix the query accordingly to your suggestion and release a new version (probably today evening or tomorrow).

Kind regards,

zepich

**Comment by zepich** [comment]
Hi @Tobi77 

I've used the fix you provided to add a new helper and replaced the query on the dashboard and in the statistic API because we used the same heavy query there as on the dashboard.

I've tested the fix with 10k+ submissions. With this amount of submissions, the full request usually needed 180 - 300ms for the dashboard and the API (on my local development machine).

When we add the statistic table in the future, we can replace the query in the helper, and both places (dashboard and statistic API) will be much faster.

I've released a new version v0.4.4 (and v1.0.0-beta.2) with this fix.

Thank you very much for your help and feedback!

Kind regards,

zepich
